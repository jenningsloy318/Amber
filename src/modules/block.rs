use std::collections::VecDeque;
use std::ops::Index;

use crate::modules::prelude::*;
use heraclitus_compiler::prelude::*;
use itertools::Itertools;
use super::statement::stmt::Statement;

#[derive(Debug, Clone)]
pub struct Block {
    pub statements: Vec<Statement>,
    pub should_indent: bool,
    pub needs_noop: bool,
    pub parses_syntax: bool,
    pub is_conditional: bool,
}

impl Block {
    // Get whether this block is empty
    pub fn is_empty(&self) -> bool {
        self.statements.is_empty()
    }

    pub fn with_condition(mut self) -> Self {
        self.is_conditional = true;
        self
    }

    pub fn with_needs_noop(mut self) -> Self {
        self.needs_noop = true;
        self
    }

    pub fn with_no_indent(mut self) -> Self {
        self.should_indent = false;
        self
    }

    pub fn with_no_syntax(mut self) -> Self {
        self.parses_syntax = false;
        self
    }
}

impl SyntaxModule<ParserMetadata> for Block {
    syntax_name!("Block");

    fn new() -> Self {
        Block {
            statements: vec![],
            should_indent: true,
            parses_syntax: true,
            needs_noop: false,
            is_conditional: false,
        }
    }

    fn parse(&mut self, meta: &mut ParserMetadata) -> SyntaxResult {
        let is_single_line = if self.parses_syntax {
            let parsed_word = token_by(meta, |word| [":", "{"].contains(&word.as_str()))?;
            parsed_word == ":"
        } else {
            false
        };

        meta.with_push_scope(|meta| {
            while meta.get_current_token().is_some() {
                // Handle the end of line
                if token(meta, "\n").is_ok() {
                    continue;
                }
                // Handle block end
                if !is_single_line && self.parses_syntax && token(meta, "}").is_ok() {
                    break;
                }
                let mut statement = Statement::new();
                if let Err(failure) = statement.parse(meta) {
                    return match failure {
                        Failure::Quiet(pos) => error_pos!(meta, pos, "Unexpected token"),
                        Failure::Loud(err) => return Err(Failure::Loud(err))
                    }
                }
                self.statements.push(statement);
                // Handle the semicolon
                token(meta, ";").ok();
                // Handle single line
                if is_single_line {
                    break;
                }
            }
            Ok(())
        })
    }
}

impl TranslateModule for Block {
    fn translate(&self, meta: &mut TranslateMetadata) -> FragmentKind {
        // Save the current statement queue and create a new one
        let mut new_queue = VecDeque::new();
        std::mem::swap(&mut meta.stmt_queue, &mut new_queue);
        let result = {
            let mut statements = vec![];
            for statement in &self.statements {
                let statement = statement.translate(meta);
                statements.extend(meta.stmt_queue.drain(..));
                statements.push(statement);
            }
            BlockFragment::new(statements, self.should_indent)
                .with_needs_noop(self.needs_noop)
                .with_condition(self.is_conditional)
                .to_frag()
        };
        // Restore the old statement queue
        std::mem::swap(&mut meta.stmt_queue, &mut new_queue);
        result
    }
}

impl DocumentationModule for Block {
    fn document(&self, meta: &ParserMetadata) -> String {
        let indices = self.statements.iter()
            .enumerate()
            .map(|(index, statement)| (index, statement.get_docs_item_name()))
            .filter_map(|(index, name)| name.map(|n| (n, index)))
            .sorted()
            .collect::<Vec<_>>();
        indices.iter()
            .map(|(_, index)| self.statements.index(*index))
            .map(|statement| statement.document(meta))
            .collect::<Vec<_>>()
            .join("")
    }
}
