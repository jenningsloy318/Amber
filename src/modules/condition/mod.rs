pub mod ifcond;
pub mod ifchain;
pub mod failed;
pub mod succeeded;
