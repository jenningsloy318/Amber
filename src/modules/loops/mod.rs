pub mod infinite_loop;
pub mod iter_loop;
pub mod while_loop;
pub mod break_stmt;
pub mod continue_stmt;
