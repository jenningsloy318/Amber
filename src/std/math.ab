/// Sums an array's contents
#[allow_absurd_cast]
pub fun math_sum(list) {
    let sum = 0
    for item in list: sum += item
    return sum
}

/// Returns a number, rounded to the nearest integer
#[allow_absurd_cast]
pub fun math_round(number: Num): Int {
    if number > 0 {
        return trust $ echo "({number}+0.5)/1" | bc $ as Int
    }

    return trust $ echo "({number}-0.5)/1" | bc $ as Int
}

/// Returns the largest integer less than or equal to a number
#[allow_absurd_cast]
pub fun math_floor(number: Num): Int {
    return trust $ echo "{number}" | awk '\{printf "%d", (\$1 < 0 ? int(\$1) - 1 : int(\$1))}' $ as Int
}

/// Returns the smallest integer greater than or equal to a number
#[allow_absurd_cast]
pub fun math_ceil(number: Num): Int {
    return math_floor(number) + 1
}

/// Returns the absolute value of a number
#[allow_absurd_cast]
pub fun math_abs(number) {
    if number < 0: return -number
    return number
}
