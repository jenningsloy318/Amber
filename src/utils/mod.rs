pub mod cc_flags;
pub mod context;
pub mod function_cache;
pub mod function_interface;
pub mod function_metadata;
pub mod import_cache;
pub mod metadata;

pub use metadata::*;

pub fn pluralize<'a>(amount: usize, single: &'a str, multiple: &'a str) -> &'a str {
    if amount > 1 {
        multiple
    } else {
        single
    }
}

/// Check if a name consists only of uppercase alphabetic characters (and optionally underscores/numbers)
pub fn is_all_caps(name: &str) -> bool {
    name.chars()
        .filter(|c| c.is_alphabetic())
        .all(|c| c.is_uppercase())
}