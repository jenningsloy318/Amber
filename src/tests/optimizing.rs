/// Tests for Amber scripts that check snapshot of generated AST.
use crate::compiler::{AmberCompiler, CompilerOptions};
use crate::modules::prelude::{FragmentRenderable, TranslateModule};
use crate::utils::TranslateMetadata;
use test_generator::test_resources;
use insta::assert_snapshot;
use std::fs;
use std::path::Path;
use crate::optimizer::optimize_fragments;

pub fn translate_and_optimize_amber_code<T: Into<String>>(code: T) -> Option<String> {
    let options = CompilerOptions::default();
    let compiler = AmberCompiler::new(code.into(), None, options);
    let tokens = compiler.tokenize().ok()?;
    let (ast, meta) = compiler.parse(tokens).ok()?;
    let mut translate_meta = TranslateMetadata::new(meta, &compiler.options);
    let mut translation = ast.translate(&mut translate_meta);
    optimize_fragments(&mut translation);
    Some(translation.to_string(&mut translate_meta))
}

/// Autoload the Amber test files in translation
#[test_resources("src/tests/optimizing/*.ab")]
fn test_translation(input: &str) {
    let code = fs::read_to_string(input)
        .unwrap_or_else(|_| panic!("Failed to open {input} test file"));
    let output = translate_and_optimize_amber_code(code)
        .expect("Couldn't translate Amber code");
    let filename = Path::new(input).file_name()
        .expect("Provided directory")
        .to_str()
        .expect("Cannot translate to string");
    assert_snapshot!(filename, output);
}
