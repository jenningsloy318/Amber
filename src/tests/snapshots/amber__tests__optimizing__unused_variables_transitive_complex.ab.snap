---
source: src/tests/optimizing.rs
expression: output
---
__1_a=10
__2_b="${__1_a}"
__3_c="${__1_a}"
if [ 1 != 0 ]; then
    __3_c="${__2_b}"
fi
echo "${__3_c}"
